sound/soc/sof/ipc4-topology.c

```diff
@@ -1399,6 +1479,86 @@ static int sof_ipc4_prepare_src_module(struct snd_sof_widget *swidget,
 	return 0;
 }
 
+static int sof_ipc4_prepare_process_module(struct snd_sof_widget *swidget,
+					   struct snd_pcm_hw_params *fe_params,
+					   struct snd_sof_platform_stream_params *platform_params,
+					   struct snd_pcm_hw_params *pipeline_params, int dir)
+{
+	struct snd_soc_component *scomp = swidget->scomp;
+	struct snd_soc_dapm_widget *widget = swidget->widget;
+	struct snd_sof_dev *sdev = snd_soc_component_get_drvdata(scomp);
+	struct sof_ipc4_process *process = swidget->private;
+	struct snd_sof_control *scontrol = NULL;
+	const struct snd_kcontrol_new *kc;
+	struct soc_bytes_ext *sbe;
+	struct sof_ipc4_control_data *control_data;
+	struct sof_ipc4_available_audio_format *available_fmt = &process->available_fmt;
+	void *cfg = process->ipc_config_data;
+	void *extra_flag;
+	int ret, i;
+
+	available_fmt->ref_audio_fmt = &available_fmt->base_config->audio_fmt;
+
+	/*
+	 * Output format is optional for process modules.
+	 * Process modules setup the output format based on audio format tokens in topology.
+	 */
+	if (swidget->output_fmt_cnt)
+		ret = sof_ipc4_init_audio_fmt(sdev, swidget, &process->base_config,
+					      &process->output_format, pipeline_params,
```

Comment:
this doesnt seem logical. If output_fmt_cnt is > 1, shouldnt you parse all output formats based on the count? Why do we only have 1 output_format in struct sof_ipc4_process?

Reply:
Actually, the output_fmt_cnt is always 0 or 1 for all our current modules. I made the output_fmt_cnt as an integer not bool just for easy to extend. There is a limitation in SOF current design. But likely, the limitation won't be hit so far. As you know, in tplg, one input_format only maps one output_format in the Object.Base.audio_format. This means even a module has 2 output pins, the output format cnt is always 1. We don't support 2 different output formats at all. This is why the output format cnt should always be 1 in the current design. And likely, this is OK as all the modules we support now do need 0 or 1 output format configuration in the init instance payload. 

If the module needs 2 or more output formats in the future, then let's talk about it at that time (I don't think we will meet such situation)